### DISCUSSION TITLE
最新版本2.2.1redis不缓存

### DISCUSSION THREAD

**Discussion by shen09** [IrineSistiana/mosdns#221] [Q&A]
为什么最新版本2.2.1的redis不缓存？我的mosdns前面套了一个adguardhome，等于adguardhome和mosdns都有缓存，adguardhome的上游dns是mosdns的，mosdns的缓存我使用的是redis，mosdns的缓存配置如下

  # 缓存
  - tag: 'mem_cache'
    type: 'cache'
    args:
      # size: 102400
      redis: 'redis://127.0.0.1:6379/0'
      lazy_cache_ttl: 86400
      lazy_cache_reply_ttl: 60

我访问了一个网页，然后重新启动adguardhome再访问同一个页面（因为现在版本的adguardhome不支持持久缓存，所以重启后 酒就自动清空了缓存），查看adguardhome的日志，通过日志里面解析的域名延迟判断mosdns没有缓存住之前已经访问过的这个域名，然后我换回了mosdns之前的那个版本同样的操作下之前访问过的域名解析延迟是1毫秒，为了测试结果的准确，每次测试我都是先清空redis的缓存，这是我的配置有问题还是什么？

**Comment by IrineSistiana** [comment]
2.2.1 版缓存小改了。只缓存有且只有一个 Question的请求的应答。其他请求不缓存。

按理说不会影响正常使用。正常用基本上所有请求都有且只有一个 Question。

前置的adh往请求里面插别的东西了？

你用 `dig +noedns google.com @mosdns的ip -p mosdns的端口` 直接测试 mosdns 的端口看看？

**Reply by shen09** [reply]
我又测试了下，好像带edns的请求都是无法缓存的，只有去除edns的请求就可以被缓存，所以没有被缓存是因为我之前adguardhome打开了edns功能，关闭之后就可以缓存了，但是关闭了edns获取到的ip地址就不是最近的了，如果我打开了edns功能，等于说缓存只使用了adguardhome的缓存，mosdns的缓存等于不起作用，另外我测试了mosdns带edns解析，好像也是不能缓存的，比如这样

dig google.com @mosdns的ip -p mosdns的端口 +subnet=111.1.1.1

所以说，edns的结果没缓存是属于正常现象吗？还是属于bug？

**Reply by IrineSistiana** [reply]
其实 mosdns 之前缓存 edns 才是 bug..... 因为 mosdns 不能按规定处理 edns，(虽然看起来能用，但其实有问题

新版修正了这个问题。所以 edns 不会缓存是正常现象。

你可以用 mosdns 自带的 ecs 插件。
